### DISCUSSION TITLE
Como o back-end acabará com o front-end

### DISCUSSION THREAD

**Discussion by HackerGaucho** [frontendbr/forum#2256] [Discussão]
O post **_[Como o back-end acabará com o front-end](https://hackergaucho.com/post/15)_**  é uma resposta para a pergunta abaixo:

<a href=https://twitter.com/VcWild/status/1498689751938965512><img src=https://i.imgur.com/aDQSVlc.png></a>

**Comment by felquis** [comment]
Oi, achei maneiro suas observações

Talvez vc possa escrever mais sobre modelos de dados (RDF) e ontologias (OWL) e ou botar links de referencias que você acredita serem as mais relevantes - pessoalmente nunca ouvi falar (até agora)

### Streaming

Similar ao Google Stadia, há um browser que também faz exatamente isso, você navega na internet mas a interface é um vídeo e suas interações são enviadas encriptadas para o server que efetua a ação e devolve para o client-side quase um UDP para o browser do que está rodando em um servidor - diz que o consumo de memória é drasticamente reduzido o que é ótimo https://www.mightyapp.com/ já podia ser o padrão que eu ficaria muito feliz.

### JSON

De fato, se só existisse JSON (no maior estilo [JAMStack](https://jamstack.org/)), a gente precisaria só se preocupar em como apresentar aquela informação para cada tecnologia de front-end, até mesmo uma hipotética AI poderia criar centenas de variações de como apresentar estes dados, o que se vê hoje é que designers em colaboração com front-enders criam centenas de templates que o usuário final pode escolher qual gosta mais.

Acredito que é o padrão principal de projetos grandes, os dados estão em um banco, uma api retorna pedaços de JSON e tanto no backend ou no client-side estes dados são interpolados com um monte de código e vira uma interface interativa de apresentação final.

E analisar o HTML de um site, transforma-lo em um JSON é algo bem divertido também, mas já ter o JSON primeiro economiza muita energia.

### Tecnologias Front-end nunca morrerão

**Algo que mudou a forma com que eu entendo o front-end**, foi uma vez que entendi que o Android e iOS nativos, tem front-end. Então, o front-end não se limita ao HTML. Se um PDF tem uma interface que apresenta dados diretamente ao consumidor final, este é o front-end, e o PDF é inclusive interativo em muitos casos.

Então, talvez o termo front-end foi popularizado por desenvolvedores HTML na Web, mas não se limita a HTML e CSS. Não sei a história e nem qual foi a primeira vez que o termo front-end foi empregado, mas duvido que o HTML esteja envolvido - enfim se alguém tiver um pouco de história neste ponto, eu gostaria de ler.

Vejo essa mudança como uma evolução/metamorfose do front-end, seja ele rodando no servidor e apenas streaming para o browser, seja ele em C# rodando um jogo em um Oculus Quest.

Ai há uma distinção entre o que é código que roda no client-side que pode ou não ter uma tela ou até mesmo um botão físico, e o que é front-end que é tudo aquilo que o consumidor final vê e pode ou não interagir com aquilo.

Essa é minha percepção e experiência não sei como é a de outras pessoas. Da pra ver que desde 2017 que comecei a ter mais interesse a este tema https://github.com/frontendbr/forum/discussions/523

**Comment by HackerGaucho** [comment]
valeu =) pretendo escrever sobre dados estruturados futuramente. no momento estou com pouco tempo disponível então só tenho respondido os comentários esclarecendo dúvidas mesmo. 

<img src=http://www.snee.com/bobdc.blog/img/rdfs.png>

tenho que estudar mais sobre o assunto antes de escrever, mas pelo que já li foi uma versão inicial do RDF para atributos serviu de base para o "[open graph](https://ogp.me/)" do facebook. ela também é usada no "[wikidata](https://www.wikidata.org/)", então a coisa é poderosa mesmo e já está em todos os lugares.

### Sobre os browsers na nuvem

que maneiro esse "mighty", não conhecia ainda. ele tem um esquema parecido com o "opera mini" que transformava tudo em WAP.

acho que esse é o **ponto central da discussão**: quanto dos sites serão modificados (estruturados) na nuvem antes de serem apresentados para o usuário. para a mim isso tende a ser diretamente proporcional as limitações de hardware dos clientes, como no caso do "opera mini" e do "google stadia".

em uma era de computação em nuvem baseada em GPUs é natural que novamente o hardware dos clientes não consigam dar conta de rodar todas as novas tecnologias da web, então novamente veremos uma repetição da história do "opera mini" como já estamos vendo.

### Sobre o conceito de front-end

sim, faltou contextualizar o termo front-end. até onde sei o termo surgiu durante a bolha da internet dos anos 2000 com a especialização cada vez maior dos antigos **webdesigners** (que chamaríamos de **full-stack** hoje em dia) houve uma divisão do trabalho e assim nasceu a **profissão de front-end**.

você que também é um dev das antigas deve lembrar que a uns 10 ou 15 anos atrás aqui no brasil ainda era comum os desenvolvedores web referirem-se a si mesmos como **webdesigners**. com a profissionalização dos devs a divisão profissional finalmente chegou até aqui também.

obviamente não estou aqui me referindo ao front-end enquanto camada de apresentação, isso já existia antes mesmo nos terminais vt100 dos anos 80 (imagem abaixo) e dos teletipos nos anos 60, embora foi apenas na era dos pcs que surgiu o **modelo de três camadas** ao qual me referi no meu tweet original ([imagem acima](https://github.com/frontendbr/forum/discussions/2256#discussion-3909395)).

<img src=https://www.linuxjournal.com/files/linuxjournal.com/linuxjournal/articles/035/3508/3508f1.jpg>

pelo meu próprio tweet você já pode perceber que falo de "**ocultar a camada de apresentação**" e não acabar com ela.

<img src=https://www.linuxjournal.com/files/linuxjournal.com/linuxjournal/articles/035/3508/3508f2.jpg>

o conceito de front-end a que me refiro no artigo é única e exclusivamente o **front-end enquanto foco de desenvolvimento web** que está descrito em detalhes [na wikipedia](https://en.wikipedia.org/wiki/Frontend_and_backend#Front-end_focused). 

<img src=https://www.linuxjournal.com/files/linuxjournal.com/linuxjournal/articles/035/3508/3508f3.jpg>

ocultando a camada de apresentação no back-end já ficaria bastante difícil para os devs front-end terem o que fazer. [estruturando os dados](https://universidadedatecnologia.com.br/dados-estruturados-e-nao-estruturados/) no back-end isso por si só já **acabaria com o front-end** dada a possibilidade de se entregar os mesmos dados estruturados no server-side diretamente no client-side.

### Front-end da web = Front-end do mobile = Front-end do desktop

com o "electron", "react" (web, mobile e desktop) além dos navegadores sem interface gráfica podemos facilmente expandir esse conceito de front-end focado na web para o front-end focado no mobile e no desktop já que atualmente **as tecnologias web, desktop e mobile parecem estar convergindo**, porém vamos parar por ai: não estou de forma alguma me referindo ao fim das **interfaces de usuário** em geral mas sim das **atividades comuns de web front-end** tal como conhecemos atualmente e a wikipedia descreve tão bem.

### Evolução divergente VS Evolução convergente

para tentar explicar mais ou menos o ocorreu em desenvolvimento web até aqui permita-me tomar emprestado dois termos usados na biologia: **evolução divergente** e **evolução convergente**.

<img src="https://3.bp.blogspot.com/-t6aQHBCVApE/WDWej5NIvtI/AAAAAAAAABo/lugXZjgoEtU7_EB9epjmXokk-gwm_qRVwCEw/s1600/analogia.jpg">

na **evolução divergente** as características biológicas tem uma origem comum porém em algum momento  da história elas divergem (homologias) ao se adaptarem a alguma nova situação, essa situação no caso do dev foi a bolha da internet dos anos 2000 que dividiu o desenvolvimento web em duas partes homólogas: front-end e back-end.

na **evolução convergente** (como o nome já diz) ocorre o contrário, as características iniciais são divergente mas ao longo do tempo tornam-se similares (analogias). obviamente não é esse o caso da divisão front-end e back-end pois elas tiveram uma origem comum no desenvolvimento web pré-bolha.

### Quando != Se != Como

de fato existe uma evolução do front-end paralela a evolução do back-end, porém quando a interface final for apenas uma interface padrão com pouca ou nenhuma customização por parte do usuário ou de devs podemos dizer sim que back-end acabou com o front-end tal qual o conhecemos. 

as tecnologias usadas no front-end vão continuar existindo pra sempre como fósseis em sites como o "internet archive", porém da mesma forma que já aconteceu com o "microsoft basic", o "turbo pascal",  o "visual basic" e outras tecnologias no passado elas com o tempo se tornarão obsoletas e estranhas aos hardwares e sistemas operacionais mais modernos.

sim, eu acredito realmente que o back-end e o front-end futuramente vão voltar a ser uma coisa só. **quando** ou **se** isso vai acontecer eu não me arrisco a dizer, o que especulei no meu artigo foi unicamente **como** isso poderia acontecer.
